_posts/2024-04-26-modules-101.markdown

```diff
@@ -0,0 +1,137 @@
+
+Modules can be written in C or other languages that have C bindings.  In this article we will go through the process of building simple modules in C and Rust (using the Valkey Module Rust SDK).  This articles expects the audience to be at least somewhat familiar with C, Rust and Redis/Valkey.  
+
+## Hello World module in C
+
+If we clone [Valkey](https://github.com/valkey-io/valkey) repo we will find numerous examples in `src/modules`.  Let's create a new file `module1.c` in the same folder.  
+
+```c
+#include "../valkeymodule.h"
+
+int hello(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
+    VALKEYMODULE_NOT_USED(argv);
+    VALKEYMODULE_NOT_USED(argc);
+    return ValkeyModule_ReplyWithSimpleString(ctx, "world1");
+}
+
+int ValkeyModule_OnLoad(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
+    VALKEYMODULE_NOT_USED(argv);
+    VALKEYMODULE_NOT_USED(argc);
+    if (ValkeyModule_Init(ctx,"module1",1,VALKEYMODULE_APIVER_1) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    if (ValkeyModule_CreateCommand(ctx,"module1.hello", hello,"",0,0,0) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    return VALKEYMODULE_OK;
+}
+```
+
+All we are doing is initalizing `module1` and creating a command `hello` that returns `world1`.  Now we need to update `src/modules/Makefile`
+
+```makefile
+all: ... module1.so
+
+module1.xo: ../valkeymodule.h
+
+module1.so: module1.xo
+	$(LD) -o $@ $^ $(SHOBJ_LDFLAGS) $(LIBS) -lc
+```
+
+Run `make` inside `src/modules` folder.  We will see numerous *.so and *.xo files created in `src/modules` including `module1.so`.  
+
+
+## Hello World module in Rust
+
+If we clone the [valkeymodule-rs](https://github.com/valkey-io/valkeymodule-rs) repo we will find `examples` folder with numerous .rs files.  Valkeymodule-rs is a recent fork of redismiodule-rs and there are still references to Redis in it.  We plan to update this repo and add it to [crates.io](https://crates.io/) but for now we will use it as is.  
+
+Let's create `examples/module2.rs`
+
+```rust
+use redis_module::{redis_module, Context, RedisResult, RedisString, RedisValue};
+
+fn hello(_ctx: &Context, _args: Vec<RedisString>) -> RedisResult {
+    Ok(RedisValue::SimpleStringStatic("world2"))
+}
+
+redis_module! {
+    name: "module2",
+    version: 1,
+    allocator: (redis_module::alloc::RedisAlloc, redis_module::alloc::RedisAlloc),
+    data_types: [],
+    commands: [
+        ["module2.hello", hello, "", 0, 0, 0],
+    ],
+}
+```
+
+Rust syntax is a bit different than C but we are creating `module2` with command `hello` that returns `world2`.  Now we need to update `Cargo.toml` in the root folder.  Add this the bottom of the file:
```

Comment:
It would probably be useful to explain what this is doing, not just say that "the syntax is different".

Reply:
added details